# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/TextDirectiveCreator.h
# Commit: 1ae864cdf71c
# Full Hash: 1ae864cdf71c71ccedafa75a70420187b19b3296
# Author: Jan-Niklas Jaeschke <jjaschke@mozilla.com>
# Date: 2025-05-27 03:54:23
# Regressor Bug: 1957557
# File Overlap Count: 1
# Description:
#   Bug 1957557, part 8 - Create Text Fragments: Find all matches for the target range in the partial document. r=dom-core,farre
#   
#   Search the partial document to find all
#   other matches for the content of the target range.
#   Then, for these matches, compute and save the
# ==============================================================================

diff -r 29735d438f26 -r 1ae864cdf71c dom/base/TextDirectiveCreator.h
--- a/dom/base/TextDirectiveCreator.h	Mon May 26 08:18:28 2025 +0000
+++ b/dom/base/TextDirectiveCreator.h	Mon May 26 08:18:28 2025 +0000
@@ -112,6 +112,36 @@
    */
   virtual void CollectContextTermWordBoundaryDistances() = 0;
 
+  /**
+   * @brief Searches the document for other occurrences of the target range and
+   *        converts the results into a comparable format.
+   *
+   * This method searches the partial document from the beginning up to the
+   * target range for occurrences of the target range content.
+   * This needs to be done differently based on whether matching is exact or
+   * range-based. For exact matching, the whole text content of the target range
+   * is searched for. For range-based matching, two search runs are required:
+   * One for the minimal `start` term (ie., the first word), which ends at the
+   * beginning of the target range. And one for the minimal `end` term (ie., the
+   * last word), which starts at the beginning of the target range and ends
+   * _before_ its end.
+   * The resulting lists of matching ranges do not exclude the target range.
+   */
+  virtual Result<Ok, ErrorResult> FindAllMatchingCandidates() = 0;
+
+  /**
+   * @brief Find all occurrences of `aSearchQuery` in the partial document.
+   *
+   * This method uses `nsFind` to perform a case-insensitive search for
+   * `aSearchQuery` in the partial document from `aSearchStart` to `aSearchEnd`.
+   *
+   * @return List of `Range`s which have the case-insensitive-same content as
+   *         `aSearchQuery`.
+   */
+  Result<nsTArray<RefPtr<AbstractRange>>, ErrorResult> FindAllMatchingRanges(
+      const nsString& aSearchQuery, const RangeBoundary& aSearchStart,
+      const RangeBoundary& aSearchEnd);
+
   nsString mPrefixContent;
   nsString mPrefixFoldCaseContent;
   nsTArray<uint32_t> mPrefixWordBeginDistances;
@@ -134,6 +164,8 @@
    * `dom.text_fragments.create_text_fragment.timeout`.
    */
   TimeoutWatchdog mWatchdog;
+
+  nsContentUtils::NodeIndexCache mNodeIndexCache;
 };
 
 /**
@@ -148,11 +180,22 @@
 
   void CollectContextTermWordBoundaryDistances() override;
 
+  Result<Ok, ErrorResult> FindAllMatchingCandidates() override;
+
+  void FindStartMatchCommonSubstringLengths(
+      const nsTArray<RefPtr<AbstractRange>>& aMatchRanges);
+
+  void FindEndMatchCommonSubstringLengths(
+      const nsTArray<RefPtr<AbstractRange>>& aMatchRanges);
+
   nsString mEndContent;
   nsString mEndFoldCaseContent;
 
   nsTArray<uint32_t> mStartWordEndDistances;
   nsTArray<uint32_t> mEndWordBeginDistances;
+
+  nsTArray<std::tuple<uint32_t, uint32_t>> mStartMatchCommonSubstringLengths;
+  nsTArray<std::tuple<uint32_t, uint32_t>> mEndMatchCommonSubstringLengths;
 };
 
 /**
@@ -166,6 +209,13 @@
   Result<Ok, ErrorResult> CollectContextTerms() override;
 
   void CollectContextTermWordBoundaryDistances() override;
+
+  Result<Ok, ErrorResult> FindAllMatchingCandidates() override;
+
+  void FindCommonSubstringLengths(
+      const nsTArray<RefPtr<AbstractRange>>& aMatchRanges);
+
+  nsTArray<std::tuple<uint32_t, uint32_t>> mCommonSubstringLengths;
 };
 }  // namespace mozilla::dom
 #endif